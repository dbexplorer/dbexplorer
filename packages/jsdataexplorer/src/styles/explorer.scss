.jsde-explorer {
    font-family: sans-serif;
    display: flex;
    gap: 16px;
    height: 100%;
}
.jsde-panel {
    border: 1px solid;
    position: relative;
    overflow-x: clip;
    height: 100%;
    display: flex;
    flex-direction: column;
    width: min-content;
}
.jsde-panel-body {
    overflow-y: auto;
    flex-grow: 1;
    height: 0;
    display: flex;
}
.jsde-panel__close {
    position: absolute;
    margin: 4px;
    width: 24px;
    height: 24px;
    right: 0;
    top: 0;
    padding: 0;
    display:flex;
    justify-content: center;
    align-items: center;
    &::after {
        content: '\2715';
    }
    z-index: 1;
}
.jsde-panel__header {
    margin: 4px;
    height: 24px;
    font-size: 24px;
    position: sticky;
    top: 0;
}