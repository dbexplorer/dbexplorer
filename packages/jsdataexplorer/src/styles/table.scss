.jsde-table {
    height: 0;
}

.jsde-table-head {
    color: white;
    background-color: black;
    position: sticky;
    top: 0;
}

.jsde-table-row {
    &:hover {
        background-color: gray;
    }
}

.jsde-table-row--explored {
    background-color: lightgrey;
}

.jsde-table-cell {
    padding: 8px;
    white-space: nowrap;
}