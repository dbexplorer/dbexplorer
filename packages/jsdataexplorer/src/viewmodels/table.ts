import { IGetDataOptions } from "../data";
import { cssPrefix, keysEqual } from "../utils";

export class TableCellViewModel {
  private text: string;
  getText() {
    return this.text;
  }
  setText(text: string) {
    this.text = text;
    this.updateCallback(this.text);
  }
  constructor(text: string = null) {
    this.text = text;
  }
  css() {
    return { root: cssPrefix("table-cell") };
  }
  updateCallback: (data: string) => any = () => { };
}
export class TableRowViewModel {
  private cells: TableCellViewModel[];
  private explored: boolean;
  getKey() {
    return this.key as string;
  }
  getCells() {
    return this.cells;
  }
  setCellsText(data: string[]) {
    if (!this.cells) {
      this.cells = data.map((s) => new TableCellViewModel(s));
    }
    else {
      data.map((s, idx) => this.cells[idx].setText(s));
    }
  }
  setExplored(explored: boolean, needUpdate: boolean = true) {
    this.explored = explored;
    if (needUpdate) this.updateCssCallback(this.css().root);
  }

  cssModifiers() {
    return {
      root: {
        explored: "--explored"
      }
    };
  }
  css() {
    return { root: cssPrefix("table-row") + (this.explored ? " " + cssPrefix("table-row") + this.cssModifiers().root.explored : "") };
  }
  constructor(data: string[], private key: string | string[]) {
    this.setCellsText(data);
  }
  updateDataCallback: (data: TableCellViewModel[]) => any;
  updateCssCallback: (css: string) => any;
  exploreCallback: () => void;
}

export class TableHeaderViewModel {
  public captionsViewModel: TableRowViewModel;
  constructor(captions: string[]) {
    this.captionsViewModel = new TableRowViewModel(captions, null);
  }
}

export class TableColumnDescription {
  name: string;
  title?: string;
  hasReference?: boolean;
  isMain?: boolean;
}

export class TableViewModel {
  private columns: TableColumnDescription[];
  private dataOffset: number = 0;
  private dataBackOffset: number = 0;
  private exploredRow: TableRowViewModel = null;
  private loadBackVisible = true;
  dataPartRowCount: number = 10;
  headerViewModel: TableHeaderViewModel;
  private setLoadBackVisible(value: boolean) {
    this.loadBackVisible = value;
    this.loadBackVisibleCallback(false);
  }
  loadData(back = false) {
    const options = { limit: this.dataPartRowCount, offset: back ? this.dataBackOffset : this.dataOffset, key: this.key, back: back };

    if (back && !this.key) {
      this.setLoadBackVisible(false);
      return;
    }

    this.getDataCallback(
      options,
      (data: any) => {
        if (back) {
          this.dataBackOffset += this.dataPartRowCount;
        }
        else {
          this.dataOffset += this.dataPartRowCount;
        }
        if (data.length > 0) {
          const rows = data.map(
            (rowData: any) => {
              const row = new TableRowViewModel(
                this.columns.map((col) => rowData.data[col.name].toString()),
                rowData.key
              )
              if (keysEqual(rowData.key, this.key)) {
                this.exploredRow = row;
                this.exploredRow.setExplored(true, false);
                this.exploreRowCallback(row);
              }
              row.exploreCallback = () => {
                this.exploredRow?.setExplored(false);
                const newExploredRow = (this.exploredRow !== row) ? row : null;
                this.exploredRow = newExploredRow;
                this.exploredRow?.setExplored(true);
                this.exploreRowCallback(newExploredRow);
              }
              return row;
            }
          )
          if (back) {
            this.rows = rows.reverse().concat(this.rows);
          }
          else {
            this.rows = this.rows.concat(rows);
          }
          this.addRowsCallback(this.rows);
        }
        else {
          if (back) this.setLoadBackVisible(false); else this.loadMoreVisibleCallback(false);
        }
        if (!this.key) this.loadBackVisibleCallback(false);
      }
    );
  }
  rows: TableRowViewModel[];
  css() {
    return {
      root: cssPrefix("table"),
      head: cssPrefix("table-head"),
      body: cssPrefix("table-body"),
      foot: cssPrefix("table-foot"),
      loadingTop: cssPrefix("table-loading-top"),
      loadingBottom: cssPrefix("table-loading-bottom"),
    };
  }
  getDataCallback: (options: IGetDataOptions, ready: any) => void;
  constructor(columns: TableColumnDescription[], private key: any = undefined, private title = "Table") {
    this.columns = columns;
    this.headerViewModel = new TableHeaderViewModel(
      columns.map((col) => col.title || col.name)
    );
    this.rows = [];
  }
  getKey() {
    return this.key;
  }
  getKeyAsString() {
    return this.key.toString();
  }
  getTitle() {
    return this.title;
  }
  getCurrentRowIndex() {
    let index = this.rows.indexOf(this.rows.filter(r => keysEqual(r.getKey(), this.key))[0]);
    if (index >= 0 && this.loadBackVisible) index++;
    return index;
  }
  getLoadBackVisible(): any {
    return true;
  }
  getLoadMoreVisible(): any {
    return true;
  }
  addRowsCallback: (rows: TableRowViewModel[]) => any;
  loadMoreVisibleCallback: (visible: boolean) => any;
  loadBackVisibleCallback: (visible: boolean) => any = () => { };
  exploreRowCallback: (row: TableRowViewModel) => any;
}
