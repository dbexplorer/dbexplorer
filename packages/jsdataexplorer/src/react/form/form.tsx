import React, { useEffect } from 'react';
import { FormViewModel } from '../../viewmodels/form';
import { FormInputField } from './field';
import { FormRelationship } from './relationship';

export function Form({ model }: { model: FormViewModel }) {
  const css = model.css();
  useEffect(() => {
    model.reloadData();
  }, [])
  return (
    <form className={css.root} action="#">
      {model.fields.map((f, index) => <FormInputField key={index} model={f} />)}
      {model.rels.map((r, index) => <FormRelationship key={index} model={r} />)}
    </form>
  );
}