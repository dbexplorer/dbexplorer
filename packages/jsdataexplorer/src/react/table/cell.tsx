import React, { useState, useEffect } from 'react';
import { TableCellViewModel } from '../../viewmodels/table';

export function TableCell({ model }: { model: TableCellViewModel }) {
  const [text, setText] = useState(model.getText());

  model.updateCallback = setText;

  return <td className={model.css().root}>{text}</td>;
}