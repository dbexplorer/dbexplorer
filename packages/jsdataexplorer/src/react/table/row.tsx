import React, { useState } from 'react';
import { TableRowViewModel } from '../../viewmodels/table';
import { TableCell } from './cell';

export function TableRow({ model }: { model: TableRowViewModel }) {
  const [rootCss, setRootCss] = useState(model.css().root);

  model.updateCssCallback = setRootCss;
  function handleClick() {
    return model.exploreCallback();
  }
  return (
    <tr className={rootCss} onClick={handleClick}>{model.getCells().map((cell, index) => <TableCell key={index} model={cell} />)}</tr>
  );
}