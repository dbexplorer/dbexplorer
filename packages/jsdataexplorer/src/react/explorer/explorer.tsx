import React, { useState, useEffect } from 'react';
import { ExplorerViewModel } from '../../viewmodels/explorer';
import { ExplorerPanel } from '../explorer/panel';

export function Explorer({ model }: { model: ExplorerViewModel }) {
  const [panels, setPanels] = useState(model.getPanels());
  useEffect(() => {
    model.start();
  }, [])

  model.addPanelCallback = () => {
    setPanels([...model.getPanels()]);
  };
  model.removePanelCallback = () => {
    setPanels([...model.getPanels()]);
  };
  const css = model.css();
  return (
    <div className={css.root}>
      {panels.map((panel, index) => <ExplorerPanel key={panel.getKey()} model={panel} />)}
    </div>
  );
}