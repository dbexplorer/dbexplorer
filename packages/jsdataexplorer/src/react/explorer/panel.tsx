import React, { useState } from 'react';
import { IBaseViewModel } from '../../viewmodels/base';
import { ExplorerPanelViewModel } from '../../viewmodels/explorer';
import { FormViewModel } from '../../viewmodels/form';
import { TableViewModel } from '../../viewmodels/table';
import { Form } from '../form/form';
import { Table } from '../table/table';

export function ExplorerPanel({ model }: { model: ExplorerPanelViewModel }) {
  const [extraDataKey, setExtra] = useState("");

  model.setExtraDataKeyCallback = setExtra;
  const css = model.css();


  function handleCloseClick(e: any) {
    model.closeCallback();
  }

  return (
    <div className={css.root}>
      <button className={css.close} onClick={handleCloseClick}></button>
      <div className={css.header}>{model.dataViewModel.getTitle()}</div>
      <div className={css.body}>
        {<Table key={model.getKey()} model={model.dataViewModel as TableViewModel}></Table>}
        {extraDataKey ? <Form key={extraDataKey} model={model.extraDataViewModel as FormViewModel}></Form> : null}
      </div>
    </div>
  );
}