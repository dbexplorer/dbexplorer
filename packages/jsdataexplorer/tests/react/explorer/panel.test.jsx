import React from 'react';
import { TableRowViewModel, TableViewModel } from "../../../src/viewmodels/table";
import { ExplorerPanelViewModel } from "../../../src/viewmodels/explorer";
import { ExplorerPanel } from "../../../src/react/explorer/panel"
import { FormViewModel } from "../../../src/viewmodels/form";
import { render, fireEvent } from '@testing-library/react'
import { act } from 'react-dom/test-utils';

jest.mock("../../../src/react/table/table", () => ({
  Table: () => {
    return <mock-table />;
  }
}));
jest.mock("../../../src/react/form/form", () => ({
  Form: () => {
    return <mock-form />;
  }
}));
test("Table panel test", () => {
  var tableViewModel = new TableViewModel([]);
  var panelViewModel = new ExplorerPanelViewModel(tableViewModel);
  tableViewModel.getDataCallback = (options, ready) => {
    ready([]);
  };
  const { container } = render(<ExplorerPanel model={panelViewModel} />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Table + form panel test", () => {
  var tableViewModel = new TableViewModel([]);
  var panelViewModel = new ExplorerPanelViewModel(tableViewModel);

  const { container } = render(<ExplorerPanel model={panelViewModel} />);

  act(() => {
    panelViewModel.extraDataViewModel = new FormViewModel([]);
    panelViewModel.setExtraDataKeyCallback("123");

    tableViewModel.getDataCallback = (options, ready) => {
      ready([]);
    };
  })
  expect(container.firstChild).toMatchSnapshot();
});

test("Close panel test", () => {
  var formViewModel = new FormViewModel([]);
  var panelViewModel = new ExplorerPanelViewModel(formViewModel);
  var d = "";
  panelViewModel.closeCallback = () => { d = "closed" };
  const { container } = render(<ExplorerPanel model={panelViewModel} />);
  fireEvent.click(container.querySelector(".jsde-panel__close"));
  expect(d).toEqual("closed");
});

test("Table panel with title test", () => {
  var tableViewModel = new TableViewModel([], undefined, "Table title");
  var panelViewModel = new ExplorerPanelViewModel(tableViewModel);
  tableViewModel.getDataCallback = (options, ready) => {
    ready([]);
  };
  const { container } = render(<ExplorerPanel model={panelViewModel} />);
  expect(container.firstChild).toMatchSnapshot();
});

test("Form panel with title test", () => {
  var formViewModel = new FormViewModel([], [], null, "Form title");
  var panelViewModel = new ExplorerPanelViewModel(formViewModel);
  formViewModel.getDataCallback = (ready) => {
    ready([{}]);
  };
  const { container } = render(<ExplorerPanel model={panelViewModel} />);
  expect(container.firstChild).toMatchSnapshot();
});