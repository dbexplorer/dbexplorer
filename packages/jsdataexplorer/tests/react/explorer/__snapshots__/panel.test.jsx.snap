// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form panel with title test 1`] = `
<div
  class="jsde-panel"
>
  <button
    class="jsde-panel__close"
  />
  <div
    class="jsde-panel__header"
  >
    Form title
  </div>
  <div
    class="jsde-panel-body"
  >
    <mock-table />
  </div>
</div>
`;

exports[`Table + form panel test 1`] = `
<div
  class="jsde-panel"
>
  <button
    class="jsde-panel__close"
  />
  <div
    class="jsde-panel__header"
  >
    Table
  </div>
  <div
    class="jsde-panel-body"
  >
    <mock-table />
    <mock-form />
  </div>
</div>
`;

exports[`Table panel test 1`] = `
<div
  class="jsde-panel"
>
  <button
    class="jsde-panel__close"
  />
  <div
    class="jsde-panel__header"
  >
    Table
  </div>
  <div
    class="jsde-panel-body"
  >
    <mock-table />
  </div>
</div>
`;

exports[`Table panel with title test 1`] = `
<div
  class="jsde-panel"
>
  <button
    class="jsde-panel__close"
  />
  <div
    class="jsde-panel__header"
  >
    Table title
  </div>
  <div
    class="jsde-panel-body"
  >
    <mock-table />
  </div>
</div>
`;
